File: src-tauri/src/model/multiplier_monitor.rs
use serde::{Deserialize, Serialize};
use sqlx::FromRow;

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct MultiplierMonitorConfig {
    pub id: String,
    pub name: String,
    pub url: String,
    pub token: String,
    pub account_ids: String,
    pub multiplier_divisor: i64,
    pub enabled: bool,
    pub last_started_at: Option<String>,
    pub last_finished_at: Option<String>,
    pub lease_owner: Option<String>,
    pub lease_until: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct MultiplierMonitorLog {
    pub id: String,
    pub run_id: String,
    pub config_id: String,
    pub config_name: String,
    pub account_id: Option<String>,
    pub account_name: Option<String>,
    pub started_at: String,
    pub finished_at: String,
    pub duration_ms: Option<i64>,
    pub result: String,
    pub old_multiplier: Option<f64>,
    pub remote_multiplier: Option<f64>,
    pub error_code: Option<String>,
    pub error_message: Option<String>,
    pub http_status: Option<i64>,
}

#[derive(Debug, Clone, FromRow)]
pub struct LatestRunStats {
    pub run_id: String,
    pub started_at: String,
    pub finished_at: String,
    pub total: i64,
    pub updated: i64,
    pub unchanged: i64,
    pub issues: i64,
    pub failed: i64,
}

File: src/api/multiplierMonitor.ts
import { get, post, put } from './client';

export type MultiplierMonitorRunSummary = {
  run_id: string;
  started_at: string;
  finished_at: string;
  total: number;
  updated: number;
  unchanged: number;
  issues: number;
  failed: number;
};

export type MultiplierMonitorConfig = {
  id: string;
  name: string;
  url: string;
  account_ids: string[];
  multiplier_divisor: number;
  enabled: boolean;
  has_token: boolean;
  has_refresh_token: boolean;
  last_started_at?: string | null;
  last_finished_at?: string | null;
  running: boolean;
  last_run?: MultiplierMonitorRunSummary | null;
  created_at: string;
  updated_at: string;
};

export type MultiplierMonitorAccountOption = {
  id: string;
  name: string;
  type: 'openai' | 'anthropic';
  status: 'active' | 'disabled';
  multiplier: number;
};

export type MultiplierMonitorCreate = {
  name: string;
  url: string;
  token: string;
  refresh_token: string;
  account_ids: string[];
  multiplier_divisor: number;
  enabled: boolean;
};

export type MultiplierMonitorUpdate = Omit<MultiplierMonitorCreate, 'token' | 'refresh_token'> & {
  token?: string;
  refresh_token?: string;
};

export type MultiplierMonitorCheckResult = {
  run_id: string;
  total: number;
  updated: number;
  unchanged: number;
  issues: number;
  failed: number;
};

export type MultiplierMonitorLog = {
  id: string;
  run_id: string;
  config_id: string;
  config_name: string;
  account_id?: string | null;
  account_name?: string | null;
  started_at: string;
  finished_at: string;
  duration_ms?: number | null;
  result: string;
  old_multiplier?: number | null;
  remote_multiplier?: number | null;
  error_code?: string | null;
  error_message?: string | null;
  http_status?: number | null;
};

export const multiplierMonitorApi = {
  list: () => get<{ items: MultiplierMonitorConfig[] }>('/api/multiplier-monitors'),
  accountOptions: () =>
    get<{ items: MultiplierMonitorAccountOption[] }>('/api/multiplier-monitors/account-options'),
  create: (payload: MultiplierMonitorCreate) =>
    post<MultiplierMonitorConfig>('/api/multiplier-monitors', payload),
  update: (id: string, payload: MultiplierMonitorUpdate) =>
    put<MultiplierMonitorConfig>(`/api/multiplier-monitors/${id}`, payload),
  toggle: (id: string) => post<MultiplierMonitorConfig>(`/api/multiplier-monitors/${id}/toggle`),
  check: (id: string) => post<MultiplierMonitorCheckResult>(`/api/multiplier-monitors/${id}/check`),
  logs: (id: string, offset: number, limit: number, result?: string) => {
    const params = new URLSearchParams({ offset: String(offset), limit: String(limit) });
    if (result) params.set('result', result);
    return get<{ items: MultiplierMonitorLog[]; total: number }>(
      `/api/multiplier-monitors/${id}/logs?${params}`,
    );
  },
};
